### PR TITLE
Update Python versions and introduce type hints

### PR DESCRIPTION
**Pull request by lewisblake**
Hi Florian,

Hope you're doing well. 

I'm opening this PR for a few things:
- to bump up the Python versions to those that are [currently active](https://devguide.python.org/versions/)
- introduce some type hinting which should help making the code a little easier to understand upon reading
- change CI to run on all branches upon pulls/pushes and also in a cron job each month

Happy for feedback!

Best,
Lewis

### CODE REVIEW

**Review by florzg — APPROVED** [review]
Thanks for the updates!

**Comment by lewisblake** [comment]
Of course! I think you may need to approve from your @florafauna account though - merging is blocked for me. Also, maybe we should tag a new version after these updates? I updated the version in `setup.py` to `v0.1.3`.

File: .github/workflows/python-package.yml
Comment (florafauna) [inline .github/workflows/python-package.yml:6]: Could you change this section to what I currently have in the repo, i.e.,
```
on:
  push:
    branches: [ "master" ]
  pull_request:
    branches: [ "master" ]
```

Needed to trigger the test pipelines...

I think the rest looks good. Also the test were running successfully:
https://github.com/florafauna/optimParallel-python/actions/runs/7267253119
(But I had to start them manually because of the missing lines above.)

I have also set up a pipeline to publish to PyPI on release now. With v0.1.3 I can test whether it works. 

Thanks.

**Review by florafauna — CHANGES_REQUESTED** [review]

File: .github/workflows/python-package.yml
Comment (lewisblake) [reply .github/workflows/python-package.yml:6]: I think the pipeline needed triggering because I was opening a PR from a fork without developer approval. I removed these lines because they're [filters](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#using-filters) to only run CI on `master`. My thought was that it would be better to run it on all branches so we know if a PR passes CI before merging in `master`. For instance, I needed to remove these lines to get CI to trigger for me on this PR (e.g., https://github.com/lewisblake/optimParallel-python/actions/runs/7267252816/job/19800655430). I could be wrong though, so I can reintroduce it if you prefer,

**Review by florafauna — APPROVED** [review]

File: .github/workflows/python-package.yml
Comment (florafauna) [reply .github/workflows/python-package.yml:6]: You are right... looks ok.

**Comment by florafauna** [comment]
v0.1.3 released on pypi.org. thanks.

**Pull request merged**
